File: test_VH/sample1_coupler.f90
program sample_coupler
  use mpi
  use ctca
  implicit none

  integer(kind=4) :: dataintnum, myrank, nprocs, ierr, areaid, progid, fromrank
  integer(kind=4),dimension(4) :: reqinfo
  integer(kind=4),dimension(2) :: dataint
  real(kind=8),dimension(6,400) :: datareal8
  integer(kind=8) :: s
  s = 600*4

  call CTCAC_init()
  print *, "coupler init done"

  call MPI_Comm_size(CTCA_subcomm, nprocs, ierr)
  call MPI_Comm_rank(CTCA_subcomm, myrank, ierr)

  do while (.true.)
     call CTCAC_pollreq_withreal8(reqinfo, fromrank, dataint, 2, datareal8, s)
  print *, "coupler poll req done"

     if (CTCAC_isfin()) then
        exit
     end if

     if (fromrank >= 0) then
  print *, "coupler got req ", dataint(2)
        ! decide progid
        progid = dataint(1)

        ! enqueue request with data
        call CTCAC_enqreq_withreal8(reqinfo, progid, dataint, 2, datareal8, s)
  print *, "coupler enqreq done"

     end if
  end do

  print *, "coupler finalize"

  call CTCAC_finalize()

end program

File: test_VH/sample1_coupler.c
#include <stdio.h>
#include <mpi.h>
#include "ctca.h"

//#define N (1024UL*1024*1024)
#define N 1024

int main()
{
    int myrank, nprocs, progid, fromrank;
    int intparams[2];
    int reqinfo[CTCAC_REQINFOITEMS];
    double data[N];
    int i, j, k;
    int prognum = 3;

    printf("coupler init\n");
    CTCAC_init_detail(10, 10, 10, N*sizeof(double), 10);
    printf("coupler init done\n");

    MPI_Comm_size(CTCA_subcomm, &nprocs);
    MPI_Comm_rank(CTCA_subcomm, &myrank);

    CTCAC_prof_start();
    CTCAC_prof_start_calc();

    while (1) {
        printf("coupler poll\n");
        CTCAC_pollreq_withreal8(reqinfo, &fromrank, intparams, 2, data, N);
        printf("coupler poll done\n");

        if (CTCAC_isfin()) 
            break;

        if (fromrank >= 0) {
        printf("coupler enq\n");
            progid = intparams[0];
            CTCAC_enqreq_withreal8(reqinfo, progid, intparams, 2, data, N);
        printf("coupler enq done\n");
        }
    }

    CTCAC_prof_stop_calc();
    CTCAC_prof_stop();

    printf("coupler fin\n");

    fprintf(stderr, "%d: coupler finalize\n", myrank);
    CTCAC_finalize();
}
